File: codewars/6kyu/longest-alphabetical-substring/dduck-longest-alphabetical-substring.js
// https://www.codewars.com/kata/longest-alphabetical-substring/javascript

const longest = (input) => {
  let longestAphabeticalString = input[0] || '';
  input.split('').reduce((acc, cur) => {
    if (cur.charCodeAt() >= acc.charCodeAt(acc.length - 1)) {
      const str = acc + cur;
      if (str.length > longestAphabeticalString.length) {
        longestAphabeticalString = str;
      }
      return str;
    }
    return cur;
  });
  return longestAphabeticalString;
};

// longest('asd');
//?
// , 'as');
longest('nab');
//?
// //, 'ab');
longest('abcdeapbcdef');
//?
// // , 'abcde');
longest('asdfaaaabbbbcttavvfffffdf');
//?
// //, 'aaaabbbbctt');
longest('asdfbyfgiklag');
//?
// // , 'fgikl');
// longest('z');
// //?
// //, 'z');
longest('zyba');
//?
//, 'z');

File: codewars/6kyu/longest-alphabetical-substring/dduck-longest-alphabetical-substring.py
# https:#www.codewars.com/kata/longest-alphabetical-substring

def longest(input):
    characters = list(map(lambda i: i, input))
    longest_aphabetical_string = alphabetical_string = characters[0] if len(characters) > 0 else ""

    for index, val in enumerate(characters):
        if ((index > 0) and (ord(characters[index]) >= ord(characters[index - 1]))):
            alphabetical_string += val
            if (len(alphabetical_string) > len(longest_aphabetical_string)):
                longest_aphabetical_string = alphabetical_string
        else:
            alphabetical_string = val

    return longest_aphabetical_string;


# print(longest('asd'))
# ?
# , 'as');
# print(longest('nab'))
# # ?
# # #, 'ab');
# print(longest('abcdeapbcdef'))
# # ?
# # # , 'abcde');
# print(longest('asdfaaaabbbbcttavvfffffdf'))
# # ?
# # #, 'aaaabbbbctt');
# print(longest('asdfbyfgiklag'))
# # ?
# # # , 'fgikl');
# # print(longest('z');
# # #?
# # #, 'z');
print(longest('zyba'))
# ?
# , 'z');
